Use bitmap Gid allocator for Gluster volumes.

humblec opened the pull request:
This patch make use of a Gid Allocator for dynamically provisioned Gluster volumes. 
A unique value is assigned from the range mentioned in the storage class.

Summary of the approach:

The Storage class has 2 more new parameters called, `gidMin and gidMax` which can be used by admin to define the GID range for the storage class.  These are optional values and if *not* filled, default values `2000 and 4294967295` are filled respectively.

The UID allocator code from openshift is used in the provisioner and we create a pool for each storage class with `block size 1`. Whenever the  DP volume create comes in, it request a new GID from the pool using "AllocateNext()" and place that as `VolumeGidAnnotationKey` in PV spec as annotation on created PV, so that POD can use it as its Supplemental GID. Whenever there is a delete request comes in , the deleter release the GID from the pool using Release() function of the allocator.

When controller is restarted, the provisioner get `all the existing PVs and filter Gluster provisioned PVs` from this list and collect all the GIDs recorded in those PVs.  Then internally make a map of
```
["storageclass1"] ["gid1", "gid2"...etc]
["storageclass2"] ["gid2"]
```
where gid1, gid2 ..etc are allocated GIDs or the GIDs of existing PVs.

After that, the provisioner create GID Pools for each storage class in above map and fill the bitmap based on previously captured GIDs.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37549)
<!-- Reviewable:end -->

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +449,18 @@ func (d *glusterfsVolumeDeleter) Delete() error {
-
 	glog.V(4).Infof("glusterfs: deleting volume %q with configuration %+v", volumeId, d.provisioningConfig)
-
+	deleteGid := d.spec.Annotations["pv.beta.kubernetes.io/gid"]
```
Use constants please, `volumehelper.VolumeGidAnnotationKey`?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +449,18 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	glog.V(4).Infof("glusterfs: deleting volume %q with configuration %+v", volumeId, d.provisioningConfig)
-
+	deleteGid := d.spec.Annotations["pv.beta.kubernetes.io/gid"]
+	scNameInSpec := d.spec.Annotations["volume.beta.kubernetes.io/storage-class"]
```
Use constants or even `GetVolumeStorageClass()`

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +449,18 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	if deleteGid64, err := strconv.ParseUint(deleteGid, 10, 32); err == nil {
+		deleteGid32 = uint32(deleteGid64)
+	}
+	releaseBlock := Block{Start: deleteGid32, End: deleteGid32}
```
You should release a GID only if `ParseUint` above succeeds.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
-
 	glog.V(4).Infof("glusterfs: creating volume with configuration %+v", r.provisioningConfig)
-	glusterfs, sizeGB, err := r.CreateVolume()
+	scName := r.options.PVC.Annotations["volume.beta.kubernetes.io/storage-class"]
```
Use constants or even GetVolumeStorageClass()

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +449,18 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	if err := gidAllocator.Release(releaseBlock); err != nil {
+		glog.Errorf("glusterfs:  error when releasing gid in storageclass: %s" , scNameInSpec)
+	}
+	glog.V(1).Infof("glusterfs: after deleteing gid (%#v), free port count:%d", deleteGid32, gidAllocator.Free())
```
Lower log level, v3 or even 4

jsafrane commented on pkg/volume/glusterfs/glusterfs_uid.go:
```diff
@@ -0,0 +1,151 @@
+
+/*
+Copyright 2015 The Kubernetes Authors.
```
2016

jsafrane commented on pkg/volume/glusterfs/glusterfs_uidallocator.go:
```diff
@@ -0,0 +1,157 @@
+package glusterfs
```
missing copyright header

jsafrane replied:
This should be probably made public in pkg/util/*, where OpenShift could find it (so we don't need to maintain two copies in two projects).

I'd appreciate if this file was "generic integer allocator" and we did not use terms like UID, GID or port here (also in the file name).

I miss unit tests.

This can be probably implemented in a separate patch.

jsafrane replied:
for the record, this has been fixed

jsafrane commented on pkg/volume/glusterfs/glusterfs_uidallocator.go:
```diff
@@ -0,0 +1,157 @@
+// and nil otherwise.
+func (r *Allocator) contains(block Block) (bool, uint32) {
+	return r.r.Offset(block)
+}
```
missing end of line

jsafrane replied:
for the record, this has been fixed

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +449,18 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+		deleteGid32 = uint32(deleteGid64)
+	}
+	releaseBlock := Block{Start: deleteGid32, End: deleteGid32}
+	gidAllocator := d.plugin.gidallocatormap[scNameInSpec]
```
Make sure it does not crash when `scNameInSpec` does not exists in `d.plugin.gidallocatormap`. GID should not be released in this case, a log message would be nice

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+	scName := r.options.PVC.Annotations["volume.beta.kubernetes.io/storage-class"]
+	
+	// TODO :Release GIDs
+	if startGid64, err := strconv.ParseUint(r.provisioningConfig.gidMin, 10, 32); err == nil {
```
* it should be probably `startGid32` if it's only 32 bit.
* handle the error please

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+		blockStart32 = uint32(startGid64)
+		glog.V(4).Infof("glusterfs: blockstart details %#v", blockStart32)
+	}
+	if endGid64, err := strconv.ParseUint(r.provisioningConfig.gidMax, 10, 32); err == nil {
```
handle the error

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+	// TODO :Release GIDs
+	if startGid64, err := strconv.ParseUint(r.provisioningConfig.gidMin, 10, 32); err == nil {
+		blockStart32 = uint32(startGid64)
+		glog.V(4).Infof("glusterfs: blockstart details %#v", blockStart32)
```
`blockstart details` looks weird to me and there already is `provisioning config and gid range %#v [%#v - %#v]` below, IMO you can delete this log (+the one for `blockend details`).

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+		blockEnd32 = uint32(endGid64)
+		glog.V(4).Infof("glusterfs: blockend details %#v ", blockEnd32)
+	}
+	glog.V(2).Infof("glusterfs: provisioning config and gid range %#v [%#v - %#v]", r.provisioningConfig, blockStart32, blockEnd32)
```
why hexadecimal `%#v`? We accept decimal numbers on input, I don't see any reason to have hexadecimal numbers in logs.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+	gidAllocatorp := r.plugin.gidallocatormap[scName]
+	if scName != "" {
+		if gidAllocatorp != nil {
+			glog.V(1).Infof("glusterfs: bitmap : %#v exist for %s ", gidAllocatorp, scName)
```
lower the log level, v(3) or even 4

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+				return nil, fmt.Errorf("glusterfs: gid pool creation failed: %v.", err)
+			}
+			gidAllocatorp = gidAllocPtr
+			glog.V(1).Infof("glusterfs: bitmap : %#v created  for %s ", gidAllocatorp, scName)
```
lower log level

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -682,6 +740,16 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
 			authEnabled = dstrings.ToLower(v) == "true"
+		case "gidmin":
+			cfg.gidMin = v
+			if err != nil {
```
where `err` comes from?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -682,6 +740,16 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+			}
+		case "gidmax":
+			cfg.gidMax = v
+			if err != nil {
```
where `err` comes from?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+		glog.V(4).Infof("glusterfs: blockend details %#v ", blockEnd32)
+	}
+	glog.V(2).Infof("glusterfs: provisioning config and gid range %#v [%#v - %#v]", r.provisioningConfig, blockStart32, blockEnd32)
+	gidAllocatorp := r.plugin.gidallocatormap[scName]
```
s/gidAllocatorp/gidAllocator

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+		if gidAllocatorp != nil {
+			glog.V(1).Infof("glusterfs: bitmap : %#v exist for %s ", gidAllocatorp, scName)
+		} else {
+			gidAllocPtr, err := gidPoolCreate(blockStart32, blockEnd32)
```
why two variables `gidAllocPtr` and `gidAllocatorp` for the same thing?

jsafrane replied:
I see new allocator created here, but I do not see where it is stored into `d.plugin.gidallocatormap`. Do I miss something?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+			glog.V(1).Infof("glusterfs: bitmap : %#v created  for %s ", gidAllocatorp, scName)
+		}
+	}
+	currentGidAllocPtr := gidAllocatorp
```
why third `currentGidAllocPtr` for the same thing?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -506,13 +561,15 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
 }
 
-func (p *glusterfsVolumeProvisioner) CreateVolume() (r *v1.GlusterfsVolumeSource, size int, err error) {
+func (p *glusterfsVolumeProvisioner) CreateVolume(reqGid int64) (r *v1.GlusterfsVolumeSource, size int, err error) {
```
IMO, int32 would be enough, code in `Provision()` works with 32 bits only. Or make it 64 bit everywhere, just make it consistent.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -526,7 +583,8 @@ func (p *glusterfsVolumeProvisioner) CreateVolume() (r *v1.GlusterfsVolumeSource
 		return nil, 0, fmt.Errorf("failed to create glusterfs REST client, REST server authentication failed")
 	}
-	volumeReq := &gapi.VolumeCreateRequest{Size: sz, Durability: gapi.VolumeDurabilityInfo{Type: durabilityType, Replicate: gapi.ReplicaDurability{Replica: replicaCount}}}
+	//volumeReq := &gapi.VolumeCreateRequest{Size: sz, Durability: gapi.VolumeDurabilityInfo{Type: durabilityType, Replicate: gapi.ReplicaDurability{Replica: replicaCount}}}
```
remove the comment

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +449,18 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	gidAllocator := d.plugin.gidallocatormap[scNameInSpec]
+	if err := gidAllocator.Release(releaseBlock); err != nil {
+		glog.Errorf("glusterfs:  error when releasing gid in storageclass: %s" , scNameInSpec)
+	}
```
Suggestion: when the last GID from `gidAllocator` is released, i.e. there is no PV with given StorageClass, you should also remove whole `gidAllocator` from `d.plugin.gidallocatormap` so we don't end up with thousands of empty allocators.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -17,35 +17,40 @@ limitations under the License.
-	exe  exec.Interface
+	host         volume.VolumeHost
+	exe          exec.Interface
+	gidallocatormap map[string]*Allocator
```
CamelCase please and I can see it's a map, no need to repeat it in variable name. `gidAllocators`?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -492,9 +513,43 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+		return nil, fmt.Errorf("glusterfs: gid allocation error: %v.", err)
+	}
+	reqGid := int64(gidblock.Start)
+	glog.V(1).Infof("glusterfs: assigned gid [%#v] ", reqGid)
```
lower log level (2 would be probably fine here), but please log also the PV or PVC name (`assigned gid %d to PV %s`).

liggitt commented:
didn't review the code closely, but a specifiable range with in-memory allocation seems reasonable as long as we're happy with a single writer (I think we are). If the allocator reads existing PVs created from its storage class and populates the allocator with the GIDs in use prior to allocating any new GIDs, it can rebuild state after a restart.

jsafrane commented on pkg/volume/glusterfs/glusterfs_uid.go:
```diff
@@ -0,0 +1,151 @@
+
+
+*/
+package glusterfs
```
* This should be probably made public in `pkg/util/*`, where OpenShift could find it (so we don't need to maintain two copies in two projects).

* I'd appreciate if this file was "generic integer range/block" and we did not use terms like UID, GID or port here (also in the file name).

* I miss documentation and unit tests.

This can be probably implemented in a separate patch.

obnoxxx replied:
I agree that this should be made public.
And as a separate, prereq patch.
see idea about a natural api above.

jsafrane commented on pkg/volume/glusterfs/glusterfs_uidallocator.go:
```diff
@@ -0,0 +1,157 @@
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	//"k8s.io/kubernetes/pkg/registry/service/allocator"
```
delete the line if it is not used

jsafrane commented on pkg/volume/glusterfs/glusterfs_uidallocator.go:
```diff
@@ -0,0 +1,157 @@
+	"k8s.io/kubernetes/pkg/api"
+	//"k8s.io/kubernetes/pkg/registry/service/allocator"
+	"k8s.io/kubernetes/pkg/registry/core/service/allocator"
+	//"github.com/openshift/origin/pkg/security/uid"
```
delete the line if it is not used

jsafrane reviewed (changes requested):
I checked glusterfs_uidallocator.go and glusterfs_uid.go only lightly,

jsafrane commented:
@k8s-bot ok to test

jsafrane commented:
I am not very happy with copying code from OpenShift, IMO there should be one generic "integer allocator", reused for UIDs (OpenShift), GIDs (Gluster) and ports (Kubernetes).

jsafrane commented:
In addition, I did not find any initialization of GID maps. After kube-controller restart it does not know what GIDs it has allocated in its previous life and can assign already used GIDs.

jsafrane commented:
I'll handle this PR, @brendandburns feel free to comment or even take it back if you're interested.

humblec commented:
@liggit Thanks for sharing your thought about this : `If the allocator reads existing PVs created from its storage class and populates the allocator with the GIDs in use prior to allocating any new GIDs, it can rebuild state after a restart.`,  yeah, thats the plan. I will bring that soon on this patchset.

humblec commented:
@jsafrane Thanks for your review ! I will address them soon.

humblec commented:
@rootfs not sure I got your comment, but `NewContiguousAllocationMap()` which we use here, internally returns `AllocationBitmap`. Please point out if you were mentioning something else wrt `AllocationBitmap`.

childsb commented:
Just to emphasize, this PR includes knobs to set the GID range for allocation (GOOD) but is missing a check that the GIDs weren't previously given out.  

A check for previously allocated GIDs is important piece of the security in scenarios where the controller is restarted since this would cause re-use of GIDs at the bottom of the range.

humblec commented:
@childsb I have tried to bring the controller restart scenario and recreation of bitmap based on previously allocated GIDs in latest patch set.  https://github.com/kubernetes/kubernetes/pull/37549/commits/cdd7cfc91058dc6681f943801e288a73bc025fb2#diff-e97253dd603331ffca81131a4b67264fR516

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,128 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	return nil
 }
 
+func (r *glusterfsVolumeProvisioner) ConstructBitmap() error {
```
Bitmap should not be part of the name, since it bleeds out implementation detais of the allocator. need a better name.

humblec replied:
Done.

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,128 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+				glog.Errorf("glusterfs: gid pool creation failed %v", err)
+				return fmt.Errorf("glusterfs: gid pool creation failed: %v.", err)
+			}
+		}
```
imho this initialization should be factored out into a function to avoid code duplication below in Create()

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,128 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+		return nil, fmt.Errorf("glusterfs: gid allocation error: %v.", err)
+	}
+	reqGid := int64(gidblock.Start)
+	glog.V(2).Infof("glusterfs: assigned gid [%d] to PVC %s", reqGid, r.options.PVC.Name)
```
This is an unnatural API, imho.

I think we should create an api like this around the allocator, dealing with Single numbers not blocks that you need to apply .Start to... :

```
... init ... --> init with a range
error Reserve(number) --> reserve a given number, error if already reserved or out of range
error Release(number) --> release a reserved number, error if not reserved or out of range
error, number Allocate() --> reserve some element.
```

something like this...

liggitt replied:
We need to be able to distinguish between out-of-range errors on Release() and "I actually encountered an error releasing". Since min/max can change and put previously allocated gids outside the controlled range, if we try to release those gids later, we need to be able to ignore an out-of-range error and continue deleting the PV

liggitt replied:
a simpler interface for the provisioner/deleter to use is fine with me

> We could in theory always detect this from traversing the PVs and thereby get along without the table altogether

reads from etcd are not guaranteed consistent :-/

also, if we have multiple threads allocating gids (which we do), we need a single lock guarding allocations

humblec replied:
@liggitt @obnoxxx afaict, its not possible to change the SC when its active and its Readonly. https://github.com/kubernetes/kubernetes/pull/37549#discussion_r90389090

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -506,13 +643,26 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
-func (p *glusterfsVolumeProvisioner) CreateVolume() (r *v1.GlusterfsVolumeSource, size int, err error) {
+func (p *glusterfsVolumeProvisioner) ConvertGid(inputGid string) (outputGid uint32, err error) {
+	if inputGid32, err := strconv.ParseUint(inputGid, 10, 32); err == nil {
+		outputGid = uint32(inputGid32)
```
separate assignment and check?

liggitt commented on pkg/volume/glusterfs/glusterfs_uidallocator.go:
```diff
@@ -0,0 +1,170 @@
+		return ErrNotInRange
+	}
+
+	allocated, err := r.alloc.Allocate(int(offset))
```
I don't see how this reserves more than a single item if you don't pass start and end

liggitt commented on pkg/volume/glusterfs/glusterfs_uidallocator.go:
```diff
@@ -0,0 +1,170 @@
+		return nil
+	}
+
+	return r.alloc.Release(int(offset))
```
how does Release know how many items to release if you're not passing along both the starting and ending points?

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -21,31 +21,39 @@ import (
-	exe  exec.Interface
+	host          volume.VolumeHost
+	exe           exec.Interface
+	gidAllocators map[string]*Allocator
```
protect gidAllocators with a RWLock and only access it via the getter

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -68,6 +76,10 @@ const (
 	annGlusterUser              = "glusterfs.kubernetes.io/userid"
 	gciGlusterMountBinariesPath = "/sbin/mount.glusterfs"
+	defaultGidMin               = "2000"
+	defaultGidMax               = "4294967295"
```
make the defaults uint32, not strings

humblec replied:
Done.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -68,6 +76,10 @@ const (
 	gciGlusterMountBinariesPath = "/sbin/mount.glusterfs"
+	defaultGidMin               = "2000"
+	defaultGidMax               = "4294967295"
+	annDynamicallyProvisioned   = "pv.kubernetes.io/provisioned-by"
```
shouldn't need this annotation... we should pick up all gids on PVs related to a storage class we're responsible for allocating from, regardless of who provisioned it

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -68,6 +76,10 @@ const (
+	defaultGidMin               = "2000"
+	defaultGidMax               = "4294967295"
+	annDynamicallyProvisioned   = "pv.kubernetes.io/provisioned-by"
+	annStorageClass             = "volume.beta.kubernetes.io/storage-class"
```
shouldn't we already know the storage class we're supposed to use? how were we getting our provisioning params before?

humblec replied:
These constants are mainly used when we construct bitmap. ie, to fetch all the existing PVs and to pick up gluster provisioned PVs from it.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	for key := range scNameGids {
+		scdetails, _ := r.plugin.host.GetKubeClient().Storage().StorageClasses().Get(key)
+		if scdetails.Parameters["gidMin"] == "" {
+			scdetails.Parameters["gidMin"] = defaultGidMin
```
storageclass params can be changed... need to handle gid min/max updating after the initial load

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -21,31 +21,39 @@ import (
+	host          volume.VolumeHost
+	exe           exec.Interface
+	gidAllocators map[string]*Allocator
+	initialized   bool
```
don't need `initialized`

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -439,9 +454,25 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	glog.V(4).Infof("glusterfs: deleting volume %q with configuration %+v", volumeId, d.provisioningConfig)
-
+	deleteGid := d.spec.Annotations[volumehelper.VolumeGidAnnotationKey]
+	scNameInSpec := storageutil.GetVolumeStorageClass(d.spec)
```
already have this as `class.Name` from above

humblec replied:
Done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	return nil
+}
+
 func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
```
nevermind, a new glusterfsVolumeProvisioner is created per provision operation... I'm still not a fan of mutating the struct vars instead of passing parameters, but it's not unsafe, as long as we don't mutate anything in the shared `plugin` var

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	if err != nil {
 		return nil, err
 	}
 	r.provisioningConfig = *cfg
```
answering my own question, a new glusterfsVolumeProvisioner instance is constructed per PVC provision operation

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	var err error
+	// Controller restart is handled here with initialized bool in plugin struct.
+	if !r.plugin.initialized {
+		err = r.ConstructBitmap()
```
the case of updating a min/max on a storage class requires some thought... there's a race in the pseudocode above:

1. provisioner A gets an allocator for sc=foo,min=1,max=1000
2. provisioner A asks the allocator for the next free value, gets gid=2
3. storage class is edited to min=1,max=2000
4. provisioner B gets an allocator for sc=foo,min=1,max=2000, which rebuilds and returns a different instance
5. provisioner B asks the allocator for the next free value, also gets gid=2
6. provisioner A persists a PV with gid=2
7. provisioner B persists a PV with gid=2

humblec replied:
@liggitt @obnoxxx  To summarize, at present  `ConStructAllocators` ( renamed from `ConstructBitmap`) is invoked only when there is a controller restart. In that func, we mainly do below.

(LOCK) -  ( Not available in patch yet)
1) Filter gluster PVs and fetch all the allocated GIDs in each storage class.
2) create a pool for above storage classes and record these GIDs in each of them.
(UNLOCK) -  ( Not available in patch yet)

If we `LOCK and UNLOCK` as shown above, I think the only concern here is about the Storage classes which got `deleted and recreated` (obv. with same class name) with a different `gidMin and gidMax` after we recreated the pool. As the allocator's pointer is  stored in plugin struct and whenever there is a new provision request comes in, it *only* validate the allocator existence and ignore the new gidMin/gidMax ( pool range) carried in this request. However when `AllocateNext()` is called from this new provision instance it should get an error if that GID does not belong to the pool range of existing allocator. In short, the new values of pool range is not respected if there was a `storage class recreation` and `not edit`.

humblec replied:
A single global allocator with locking is avilable here https://github.com/humblec/kubernetes/commit/0f6a8e471f07ed5585e86e0d699d4ed6f4a49f23

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+		}
+	}
+	glog.V(1).Infof("glusterfs: current gid allocator details %#+v", r.plugin.gidAllocators[scName])
+	gidblock, err := r.plugin.gidAllocators[scName].AllocateNext()
```
if there's an error creating the PV with this gid, need to release it from the allocator

humblec replied:
Done.

humblec committed:
Use bitmap Gid allocator for Gluster volumes.

The gid pool can be configured from storage class.
This also reconstruct gid allocator bitmap after controller restart.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -385,6 +397,8 @@ type provisioningConfig struct {
 	secretNamespace string
 	secretName      string
 	secretValue     string
+	gidMin          string
```
Done.

humblec commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
-	glusterfs, sizeGB, err := r.CreateVolume()
+	scName := r.options.PVC.Annotations[annStorageClass]
+	// TODO :Release GIDs
+	if startGid32, err := strconv.ParseUint(r.provisioningConfig.gidMin, 10, 32); err == nil {
```
Done.

jsafrane reviewed (commented):
all my remarks seem to be fixed, now please fix Jordan's

liggitt commented:
closing in favor of https://github.com/kubernetes/kubernetes/pull/37886
